src/OptimalFilter.jl

```diff
@@ -139,59 +139,64 @@ end
-function normalized_2d_fft!(transformed_array::AbstractMatrix{T}, array::AbstractMatrix{S}, grid_ext) where {T,S}
+function normalized_2d_fft!(transformed_array::AbstractMatrix{<:Complex}, array::AbstractMatrix{S}, fft_plan::FFTW.FFTWPlan, fft_plan!::FFTW.FFTWPlan, grid_ext, f=identity) where {S}
 
-    normalization_factor = 1.0 / sqrt(grid_ext.nx * grid_ext.ny)
-    transformed_array .= fft(array) .* normalization_factor
+    normalization_factor = f(sqrt(grid_ext.nx * grid_ext.ny))
+    if pointer(transformed_array) == pointer(array)
+        mul!(transformed_array, f(fft_plan!), array)
+    else
+        mul!(transformed_array, f(fft_plan), array)
+    end
+    transformed_array ./= normalization_factor
 
 end
 
 # Normalized two-dimension discrete Fourier transofrm normalized by sqrt(n1_bar).
 # Operates on the 2d data stored as a vector.
+# The argument `f` lets you switch between forward transform (`f=identity`) and
+# inverse transform (`f=inv`).
 # From Dietrich & Newsam 96 in text following equation 12
-function normalized_2d_fft!(transformed_vector::AbstractVector{T}, vector::AbstractVector{S}, grid_ext) where {T,S}
-
-    normalization_factor = 1.0 / sqrt(grid_ext.nx * grid_ext.ny)
-    transformed_vector .= fft(reshape(vector, grid_ext.nx, grid_ext.ny))[:] .* normalization_factor
-
-end
-
-# Normalized inverse two-dimension discrete Fourier transofrm normalized by sqrt(n1_bar).
-# Operates on the 2d data stored as a matrix.
-# From Dietrich & Newsam 96 in text following equation 12
-function normalized_inverse_2d_fft!(transformed_array::AbstractMatrix{T}, array::AbstractMatrix{S}, grid_ext) where {T,S}
+function normalized_2d_fft!(transformed_vector::AbstractVector{<:Complex}, vector::AbstractVector{S}, fft_plan::FFTW.FFTWPlan, fft_plan!::FFTW.FFTWPlan, grid_ext, f=identity) where {S}
 
-    normalization_factor = sqrt(grid_ext.nx * grid_ext.ny)
-    transformed_array .= ifft(array) .* normalization_factor
+    normalization_factor = f(sqrt(grid_ext.nx * grid_ext.ny))
+    tmp_array = complex(reshape(vector, grid_ext.nx, grid_ext.ny))
+    mul!(tmp_array, f(fft_plan!), tmp_array)
+    transformed_vector .= @view(tmp_array[:]) ./ normalization_factor
 
 end
 
 # Normalized inverse two-dimension discrete Fourier transofrm normalized by sqrt(n1_bar).
-# Operates on the 2d data stored as a vector.
 # From Dietrich & Newsam 96 in text following equation 12
-function normalized_inverse_2d_fft!(transformed_vector::AbstractVector{T}, vector::AbstractVector{S}, grid_ext) where {T,S}
+function normalized_inverse_2d_fft!(transformed_array::AbstractArray{<:Complex}, array::AbstractArray{S}, fft_plan::FFTW.FFTWPlan, fft_plan!::FFTW.FFTWPlan, grid_ext) where {S}
 
-    normalization_factor = sqrt(grid_ext.nx * grid_ext.ny)
-    transformed_vector .= ifft(reshape(vector, grid_ext.nx, grid_ext.ny))[:] .* normalization_factor
+    return normalized_2d_fft!(transformed_array, array, fft_plan, fft_plan!, grid_ext, inv)
 
 end
```

Comment:
Do we need this function if we can just call `normalized_2d_fft!` with `inv`?

Reply:
That's a good point, I'm going to remove this function